HPCC-14476 Add support for dynamic import statements to ESDL tool

rpastrana opened the pull request:
- removes hardcoded import statement
- provides mechanism for caller to provide comma delimited list of imports

Signed-off-by: rpastrana rodrigo.pastrana@lexisnexis.com

rpastrana committed:
HPCC-14476 Adds support for dynamic import statements

- removes hardcoded import statement
- provides mechanism for caller to provide comma delimited list of imports

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
This change fixes the hard-coded import statement issue, and provides a mechanism for callers to provide a list of import statements to be included.

rpastrana commented:
@afishbeck please review.
If you need the "import $.share" we should address that in a dedicated jira/git issue. This change is meant to fix the issue of the hardcoded import $.share, and provide a means to add the appropriate import statements dynamically.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

rpastrana committed:
HPCC-14476 Represent String Arrays as ECL set of string

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on esp/xslt/esdl2ecl.xslt:
```diff
@@ -17,26 +17,57 @@
+    <xsl:template match="/">
+        <xsl:apply-templates select="expesdl"/>
+    </xsl:template>
+    <xsl:template match="text()" name="outputImports">
```
Does this match do anything on this named template?

afishbeck commented on esp/xslt/esdl2ecl.xslt:
```diff
@@ -17,26 +17,57 @@
-		<xsl:text>/*** Not to be hand edited (changes will be lost on re-generation) ***/
+    <xsl:output method="text" version="1.0" encoding="UTF-8" indent="yes"/>
+    <xsl:param name="sourceFileName" select="'UNKNOWN'"/>
+    <xsl:param name="importsList" select="''"/>
```
Is there a way to cause this parameter to be set?  i.e. via the esdl tools?

rpastrana replied:
ESDL ECL now supports this xslt feature

afishbeck commented:
Looks good except minor comment, and I'd like to know how to cause the imports to be set.

afishbeck commented:
@rpastrana

afishbeck commented on esp/xslt/esdl2ecl.xslt:
```diff
@@ -17,26 +17,57 @@
+    <xsl:template match="text()" name="outputImports">
+    <xsl:param name="import" select="."/>
+        <xsl:if test="string-length($import) > 0">
+            <xsl:text>import $.</xsl:text><xsl:value-of select="$import"/><xsl:text>;
```
Actually since the import mechanism would be ECL specific maybe we should pass in the whole import, including the "$."?  Seems like that would give more flexibility.

rpastrana replied:
Added a separate way to input a free-hand ecl "header"

afishbeck replied:
Still wonder if we always want the "$.".  Afraid that if we always append it now, we won't be able to tell when not to append it later?

rpastrana replied:
Removed

rpastrana committed:
HPCC-14476 Adds support for declared ECL imports and ECL header

- Fixes associated usage verbage

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@afishbeck take a look at the latest changes

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on tools/esdlcmd/esdl2ecl.cpp:
```diff
@@ -367,15 +374,19 @@ class Esdl2EclCmd : public EsdlConvertCmd
                 "      --includes            Process all included files\n"
-                "      --rollup              Roll-up all processed includes to single ecl output file"
+                "      --rollup              Roll-up all processed includes to single ecl output file\n"
+                "      --CDE, --compfilesdir HPCC Component files directory (xslt files)\n"
```
Since compfilesdir is also abbreviated, might as well just use CDE? Also since it is an abbreviation should use only one dash (and usually would be lower case).  -cde=<directroy>

rpastrana replied:
Changed.

afishbeck commented on tools/esdlcmd/esdl2ecl.cpp:
```diff
@@ -367,15 +374,19 @@ class Esdl2EclCmd : public EsdlConvertCmd
-                "      --rollup              Roll-up all processed includes to single ecl output file"
+                "      --rollup              Roll-up all processed includes to single ecl output file\n"
+                "      --CDE, --compfilesdir HPCC Component files directory (xslt files)\n"
+                "      --ecl-includes        Coma-delimited import list to be attached to output ECL\n"
```
Should probably be --ecl-imports for consistency.

afishbeck replied:
Command line support is good, but still think a new ESDL keyword would be more user freindly so they don't have to type it each time.  ECLImport("xxxx"). for example.  Or use or modify the keyword that already exists... I know there is something similar used for the generated c++.

rpastrana replied:
changed.

Regarding ECLImport comment, it sounds like a good feature enhancement but somewhat out of the scope of this fix. This change was only intended to removed the hard-coded import $.share statement preventing ESDL ecl filename.ecm to work for all cases. Perhaps a new ticket should be opened for that feature.

afishbeck replied:
Ok, yes may need a an issue for this feature at some point.

rpastrana committed:
HPCC-14476 Further enhances simple bug fix

- Removes --compfilesdir and --CDE
- Adds --ecl-imports
- No longer prefixes generated import statements with "$."

Signed-off-by: rpastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@afishbeck please review latest requested changes, and comments.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented:
Ok, looks fine.

afishbeck commented:
@rpastrana

rpastrana commented:
@richardkchapman please merge. Low risk change isolated to ESDL tool.
